dynamicio/io/s3.py

```diff
@@ -0,0 +1,82 @@
+from functools import partial
+from pathlib import Path
+from typing import Any, Dict, List, Literal, Optional, Type
+
+import boto3
+import pandas as pd
+
+from dynamicio.io.s3_contexts import s3_named_file_reader, s3_writer, s3_reader
+from dynamicio.io.resource import BaseResource
+from dynamicio.io.serde import BaseSerde, CsvSerde, HdfSerde, JsonSerde, ParquetSerde, PickleSerde
+
+
+class S3Resource(BaseResource):
+    bucket: str
+    path: Path
+    read_kwargs: Dict[str, Any] = {}
+    write_kwargs: Dict[str, Any] = {}
+    injectables: List[str] = ["path"]
+    file_type: Optional[Literal["parquet", "hdf", "csv", "json", "pickle"]] = None
+    force_read_to_memory: bool = False
+
+    @property
+    def _s3_path(self) -> str:
+        """For logging purposes only."""
+        return f"s3://{self.bucket}/{self.path}"
+
+    def _read(self) -> pd.DataFrame:
+        if self.force_read_to_memory:
+            with s3_reader(boto3.client("s3"), s3_bucket=self.bucket, s3_key=str(self.path)) as fobj:  # type: ignore
+                df = self.get_serde()._read(fobj, **self.read_kwargs)  # type: ignore
+                if df is not None:
+                    return df
```

Comment:
nit: don't we want to just return df here? when would the result of `_read` be None? also, as it is I think that if df is None we would end up in the not `self.force_read_to_memory` branch, is this what we want? I guess this is also going to save the file locally, which is probably going against the `force_read_to_memory` option

Reply:
That's how it was done previously - if we want to improve this, I would just raise here if df is `None`.